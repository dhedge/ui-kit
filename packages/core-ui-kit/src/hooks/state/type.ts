import type {
  TradingPanelActionsState,
  TradingPanelState,
} from 'types/state.types'

import { useTradingPanelActions, useTradingPanelState } from './context'

export const useTradingPanelType = (): [
  TradingPanelState['type'],
  TradingPanelActionsState['setTradingType'],
] => [useTradingPanelState().type, useTradingPanelActions().setTradingType]

export const useIsDepositTradingPanelType = (): boolean =>
  useTradingPanelState().type === 'deposit'
