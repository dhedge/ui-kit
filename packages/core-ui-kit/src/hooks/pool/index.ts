export { useCheckWhitelist } from './use-check-whitelist'
export { useIsDhedgePool } from './use-is-dhedge-pool'
export { useManagerLogicAddress } from './use-manager-logic-address'
export { useContractPoolComposition } from './use-contract-pool-composition'
export { usePoolComposition } from './use-pool-composition'
export { usePoolDynamicContractData } from './use-pool-dynamic-contract-data'
export { usePoolManagerLogicData } from './use-pool-manager-logic-data'
export { usePoolCompositionWithFraction } from './use-pool-composition-with-fraction'
export { usePoolTokenPrice } from './use-pool-token-price'
export { usePoolFees } from './use-pool-fees'
