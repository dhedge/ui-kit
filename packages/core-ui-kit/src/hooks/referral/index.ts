export { useReferralProgram } from './use-referral-program'
