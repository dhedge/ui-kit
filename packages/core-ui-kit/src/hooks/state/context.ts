import { useContext } from 'react'

import { TradingPanelActionsContext, TradingPanelStateContext } from 'providers'

export const useTradingPanelState = () => useContext(TradingPanelStateContext)

export const useTradingPanelActions = () =>
  useContext(TradingPanelActionsContext)
