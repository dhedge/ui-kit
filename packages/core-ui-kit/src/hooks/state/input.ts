import type {
  TradingPanelActionsState,
  TradingPanelState,
} from 'types/state.types'

import { useTradingPanelActions, useTradingPanelState } from './context'

export const useTradingPanelInput = () => useTradingPanelState().input

export const useSendTokenInput = (): [
  TradingPanelState['input']['sendToken'],
  TradingPanelActionsState['updateSendTokenInput'],
] => [
  useTradingPanelInput().sendToken,
  useTradingPanelActions().updateSendTokenInput,
]

export const useReceiveTokenInput = (): [
  TradingPanelState['input']['receiveToken'],
  TradingPanelActionsState['updateReceiveTokenInput'],
] => [
  useTradingPanelInput().receiveToken,
  useTradingPanelActions().updateReceiveTokenInput,
]
