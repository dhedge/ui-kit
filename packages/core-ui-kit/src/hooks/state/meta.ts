import { DEFAULT_DEPOSIT_LOCKTIME_MAP, DEFAULT_DEPOSIT_METHOD } from 'const'
import { useTradingPanelPoolConfig } from 'hooks/state'
import { useCallback } from 'react'
import type {
  TradingPanelActionsState,
  TradingPanelState,
} from 'types/state.types'

import { useUpdateEntryFee } from './action'
import { useTradingPanelActions, useTradingPanelState } from './context'

export const useTradingPanelMeta = (): [
  TradingPanelState['meta'],
  TradingPanelActionsState['updateTradingMeta'],
] => [useTradingPanelState().meta, useTradingPanelActions().updateTradingMeta]
export const useTradingPanelApprovingStatus = (): [
  TradingPanelState['meta']['approvingStatus'],
  (approvingStatus: TradingPanelState['meta']['approvingStatus']) => void,
] => {
  const [meta, updateMeta] = useTradingPanelMeta()

  const setApprovingStatus = useCallback(
    (approvingStatus: TradingPanelState['meta']['approvingStatus']) =>
      updateMeta({ approvingStatus }),
    [updateMeta],
  )

  return [meta.approvingStatus, setApprovingStatus]
}

export const useTradingPanelEntryFee = (): [
  number,
  (entryFeeSlice: Partial<TradingPanelState['entryFee']>) => void,
] => {
  const { depositParams } = useTradingPanelPoolConfig()
  const entryFee = useTradingPanelState().entryFee
  const updateEntryFee = useUpdateEntryFee()

  return [
    entryFee[depositParams.method ?? DEFAULT_DEPOSIT_METHOD],
    updateEntryFee,
  ]
}

export const useTradingPanelLockTime = (): string => {
  const { depositParams } = useTradingPanelPoolConfig()

  return DEFAULT_DEPOSIT_LOCKTIME_MAP[
    depositParams.method ?? DEFAULT_DEPOSIT_METHOD
  ]
}
