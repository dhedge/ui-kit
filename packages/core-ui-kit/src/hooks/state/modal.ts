import type {
  TradingPanelActionsState,
  TradingPanelState,
} from 'types/state.types'

import { useTradingPanelActions, useTradingPanelState } from './context'

export const useTradingPanelModal = (): [
  TradingPanelState['modal'],
  TradingPanelActionsState['updateTradingModal'],
] => [useTradingPanelState().modal, useTradingPanelActions().updateTradingModal]
