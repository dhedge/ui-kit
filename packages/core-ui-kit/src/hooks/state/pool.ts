import { useMemo } from 'react'

import { DEFAULT_DEPOSIT_METHOD, EMPTY_POOL_CONFIG } from 'const'
import { useUpdatePoolFallbackData } from 'hooks/state'
import type { PoolConfig } from 'types/config.types'
import type {
  TradingPanelActionsState,
  TradingPanelState,
} from 'types/state.types'
import type { DepositMethodName } from 'types/trading-panel.types'
import type { Address } from 'types/web3.types'

import { useTradingPanelActions, useTradingPanelState } from './context'

export const useTradingPanelPoolAddress = (): [
  TradingPanelState['poolAddress'],
  TradingPanelActionsState['setPoolAddress'],
] => [
  useTradingPanelState().poolAddress,
  useTradingPanelActions().setPoolAddress,
]

export const useTradingPanelPoolConfig = () => {
  const { poolAddress, poolConfigMap } = useTradingPanelState()
  return poolConfigMap[poolAddress] ?? EMPTY_POOL_CONFIG
}

export const useTradingPanelPoolConfigs = () => {
  const { poolConfigMap } = useTradingPanelState()
  return useMemo(() => Object.values(poolConfigMap), [poolConfigMap])
}

export const useIsPoolAddress = (address: Address) => {
  const configs = useTradingPanelPoolConfigs()
  return useMemo(
    () => configs.some((config) => config.address === address),
    [configs, address],
  )
}

export const useTradingPanelDepositMethod = (): [
  DepositMethodName,
  (
    payload: Pick<PoolConfig, 'address'> &
      Pick<PoolConfig['depositParams'], 'method'>,
  ) => void,
] => {
  const {
    depositParams: { method: depositMethod = DEFAULT_DEPOSIT_METHOD },
  } = useTradingPanelPoolConfig()
  const { updatePoolConfigDepositMethod } = useTradingPanelActions()

  return [depositMethod, updatePoolConfigDepositMethod]
}

export const useTradingPanelPoolFallbackData = (): [
  TradingPanelState['poolFallbackData'],
  TradingPanelActionsState['updatePoolFallbackData'],
] => {
  const { poolFallbackData, poolAddress } = useTradingPanelState()
  const updatePoolFallback = useUpdatePoolFallbackData()

  return [
    poolFallbackData.address === poolAddress
      ? poolFallbackData
      : { address: poolAddress },
    updatePoolFallback,
  ]
}
