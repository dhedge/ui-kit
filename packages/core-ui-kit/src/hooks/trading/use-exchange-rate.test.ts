import { DEFAULT_PRECISION } from 'const'
import * as stateHooks from 'hooks/state'
import { renderHook } from 'test-utils'

import { TEST_ADDRESS } from 'tests/mocks'
import type { TradingToken } from 'types'

import { getExchange, useExchangeRate } from './use-exchange-rate'

vi.mock('hooks/state', () => ({
  useReceiveTokenInput: vi.fn(),
  useSendTokenInput: vi.fn(),
}))

describe('useExchangeRate', () => {
  it('should evaluate value from sendToken and receiveToken', () => {
    const sendToken: TradingToken = {
      address: TEST_ADDRESS,
      symbol: 'send_symbol',
      value: '1',
      decimals: DEFAULT_PRECISION,
    }

    const receiveToken: TradingToken = {
      address: TEST_ADDRESS,
      symbol: 'receive_symbol',
      value: '2',
      decimals: DEFAULT_PRECISION,
    }

    vi.mocked(stateHooks.useSendTokenInput).mockImplementation(() => [
      sendToken,
      vi.fn(),
    ])
    vi.mocked(stateHooks.useReceiveTokenInput).mockImplementation(() => [
      receiveToken,
      vi.fn(),
    ])

    const { result } = renderHook(() => useExchangeRate())

    expect(vi.mocked(stateHooks.useSendTokenInput)).toHaveBeenCalledTimes(1)
    expect(vi.mocked(stateHooks.useReceiveTokenInput)).toHaveBeenCalledTimes(1)
    expect(result.current).toMatchSnapshot()
  })
})

describe('getExchange', () => {
  it('should build 1 X symbol = n Y symbol pattern string', () => {
    const input = 10
    const output = 20
    expect(
      getExchange(
        'sendSymbol',
        'receiveSymbol',
        input.toString(),
        output.toString(),
      ),
    ).toEqual(`1 sendSymbol = ${(output / input).toFixed(6)} receiveSymbol`)
  })

  it('should return empty string on invalid data', () => {
    const input = Infinity
    const output = 20
    expect(
      getExchange(
        'sendSymbol',
        'receiveSymbol',
        input.toString(),
        output.toString(),
      ),
    ).toEqual('')
  })

  it('should return empty string on zero input', () => {
    const input = 0
    const output = 20
    expect(
      getExchange(
        'sendSymbol',
        'receiveSymbol',
        input.toString(),
        output.toString(),
      ),
    ).toEqual('')
  })
})
