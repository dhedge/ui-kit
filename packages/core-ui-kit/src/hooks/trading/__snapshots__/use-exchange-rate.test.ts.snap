// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useExchangeRate > should evaluate value from sendToken and receiveToken 1`] = `
{
  "isLoading": false,
  "value": "1 send_symbol = 2.000000 receive_symbol",
}
`;
